File: src/jsonrpc/eth/getFilterLogs/eth_getFilterLogs.zig
const std = @import("std");
const types = @import("../../types.zig");

/// Returns an array of all logs matching the filter with the given ID (created using `eth_newFilter`).
///
/// Example:
/// Filter identifier: "0x01"
/// Result: ...
///
/// Implements the `eth_getFilterLogs` JSON-RPC method.
pub const EthGetFilterLogs = @This();

/// The JSON-RPC method name
pub const method = "eth_getFilterLogs";

/// Parameters for `eth_getFilterLogs`
pub const Params = struct {
    /// hex encoded unsigned integer
    filter identifier: types.Quantity,

    pub fn jsonStringify(self: Params, jws: *std.json.Stringify) !void {
        try jws.beginArray();
        try jws.write(self.filter identifier);
        try jws.endArray();
    }

    pub fn jsonParseFromValue(allocator: std.mem.Allocator, source: std.json.Value, options: std.json.ParseOptions) !Params {
        if (source != .array) return error.UnexpectedToken;
        if (source.array.items.len != 1) return error.InvalidParamCount;

        return Params{
            .filter identifier = try std.json.innerParseFromValue(types.Quantity, allocator, source.array.items[0], options),
        };
    }
};

/// Result for `eth_getFilterLogs`
pub const Result = struct {
    /// Filter results
    value: types.Quantity,

    pub fn jsonStringify(self: Result, jws: *std.json.Stringify) !void {
        try jws.write(self.value);
    }

    pub fn jsonParseFromValue(allocator: std.mem.Allocator, source: std.json.Value, options: std.json.ParseOptions) !Result {
        return Result{
            .value = try std.json.innerParseFromValue(types.Quantity, allocator, source, options),
        };
    }
};

File: src/jsonrpc/eth/getFilterLogs/eth_getFilterLogs.js
/**
 * @fileoverview eth_getFilterLogs JSON-RPC method
 */

import { createRequest } from "../../types/JsonRpcRequest.js";

/**
 * @typedef {import('../../types/index.js').Address} Address
 * @typedef {import('../../types/index.js').Hash} Hash
 * @typedef {import('../../types/index.js').Quantity} Quantity
 * @typedef {import('../../types/index.js').BlockTag} BlockTag
 * @typedef {import('../../types/index.js').BlockSpec} BlockSpec
 * @typedef {import('../../types/JsonRpcRequest.js').JsonRpcRequest} JsonRpcRequest
 */

/**
 * Returns an array of all logs matching the filter with the given ID (created using `eth_newFilter`).
 *
 * @example
 * Filter identifier: "0x01"
 * Result: ...
 *
 * Implements the `eth_getFilterLogs` JSON-RPC method.
 */

/** The JSON-RPC method name */
export const method = "eth_getFilterLogs";
/**
 * Result for `eth_getFilterLogs`
 *
 * Filter results
 *
 * @typedef {Quantity} Result
 */

/**
 * Request for `eth_getFilterLogs`
 *
 * @typedef {JsonRpcRequest<'eth_getFilterLogs', [Quantity]>} Request
 */

/**
 * Creates a eth_getFilterLogs JSON-RPC request
 *
 * @param {Quantity} address
 * @param {number | string | null} [id] - Optional request ID
 * @returns {Request}
 */
export function GetFilterLogsRequest(address, id = null) {
	return /** @type {Request} */ (createRequest(method, [address], id));
}
